crates/uv-auth/src/middleware.rs

```diff
@@ -40,102 +75,497 @@ impl Middleware for AuthMiddleware {
         extensions: &mut Extensions,
         next: Next<'_>,
     ) -> reqwest_middleware::Result<Response> {
-        let url = request.url().clone();
-
-        // If the request already has an authorization header, we don't need to do anything.
-        // This gives in-URL credentials precedence over the netrc file.
-        if request
-            .headers()
-            .contains_key(reqwest::header::AUTHORIZATION)
-        {
-            debug!("Credentials are already present for {url}");
-            return next.run(request, extensions).await;
-        }
+        // Check for credentials attached to (1) the request itself
+        let credentials = Credentials::from_request(&request);
+        // In the middleware, existing credentials are already moved from the URL
+        // to the headers so for display purposes we restore some information
+        let url = if tracing::enabled!(tracing::Level::DEBUG) {
+            let mut url = request.url().clone();
+            credentials
+                .as_ref()
+                .and_then(|credentials| credentials.username())
+                .and_then(|username| url.set_username(username).ok());
```

Comment:
I think we should avoid putting effects in optional chains like this. I'd suggest something like:

```rust
if let Some(username) = credentials.as_ref().and_then(...) {
  let _ = url.set_username(username);
}
```

It took me a bit to see that this was mutating the URL.

Reply:
Okay no problem, just felt verbose.